### PR TITLE
KAFKA-14784: Connect offset reset REST API

### PR DESCRIPTION
**Pull request by yashmayya**
- https://issues.apache.org/jira/browse/KAFKA-14784
- [KIP-875: First-class offsets support in Kafka Connect](https://cwiki.apache.org/confluence/display/KAFKA/KIP-875%3A+First-class+offsets+support+in+Kafka+Connect)
- Implements the new `DELETE /connectors/{connector}/offsets` REST API
- Relevant PRs: https://github.com/apache/kafka/pull/13434, https://github.com/apache/kafka/pull/13424, https://github.com/apache/kafka/pull/13465

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by yashmayya** [comment]
#### Sink connector offsets alter requests:
```
ConnectorsResource::alterConnectorOffsets -> AbstractHerder::alterConnectorOffsets -> (Distributed|Standalone)Herder::modifyConnectorOffsets -> Worker::alterConnectorOffsets -> Worker::modifySinkConnectorOffsets -> Worker::alterSinkConnectorOffset
```

#### Source connector offsets alter requests:
```
ConnectorsResource::alterConnectorOffsets -> AbstractHerder::alterConnectorOffsets -> (Distributed|Standalone)Herder::modifyConnectorOffsets -> Worker::alterConnectorOffsets -> Worker::modifySourceConnectorOffsets
```

#### Sink connector offsets reset requests:
```
ConnectorsResource::resetConnectorOffsets -> AbstractHerder::resetConnectorOffsets -> (Distributed|Standalone)Herder::modifyConnectorOffsets -> Worker::resetConnectorOffsets -> Worker::modifySinkConnectorOffsets -> Worker::resetSinkConnectorOffsets
```

#### Source connector offsets reset requests:
```
ConnectorsResource::resetConnectorOffsets -> AbstractHerder::resetConnectorOffsets -> (Distributed|Standalone)Herder::modifyConnectorOffsets -> Worker::resetConnectorOffsets -> Worker::modifySourceConnectorOffsets
```

The current flows for altering and resetting offsets along with the use of `null` offsets in multiple places to distinguish between alter and reset offsets requests might seem a little clunky (especially for sink connectors), but I've tried to optimize for code re-use in both the herder implementations as well as the worker.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 1]
```diff
@@ -1320,89 +1338,188 @@ void alterSinkConnectorOffsets(String connName, Connector connector, Map<String,
...
+                        throw new ConnectException("Failed to modify offsets for connector " + connName + " because it doesn't support external " +
+                                "modification of offsets", e);
+                    }
 
-                        adminFutures.add(deleteConsumerGroupOffsetsResult.all());
+                    // This should only occur for an offset reset request when:
+                    // 1. There was a prior attempt to reset offsets
+                    // OR
+                    // 2. No offsets have been committed yet
+                    if (offsetsToWrite.isEmpty()) {
```
Comment (yashmayya) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: I'm wondering whether we should go ahead and attempt to delete the consumer group even in this case, in order to avoid inconsistency (if no offsets have been committed for the group yet, the current implementation won't delete the consumer group). The only minor downside is that we'd need special case handling for `GroupIdNotFoundException`s arising from calls to `Admin::deleteConsumerGroups` (interestingly `Admin::listConsumerGroupOffsets` doesn't result in an exception for non-existent groups, but an empty partition offset map).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java [hunk 2]
```diff
@@ -374,19 +374,38 @@ public synchronized void connectorOffsets(String connName, Callback<ConnectorOff
...
             cb.onCompletion(new NotFoundException("Connector " + connName + " not found", null), null);
-            return;
+            return false;
         }
 
         if (configState.targetState(connName) != TargetState.STOPPED || configState.taskCount(connName) != 0) {
-            cb.onCompletion(new BadRequestException("Connectors must be in the STOPPED state before their offsets can be altered. " +
-                    "This can be done for the specified connector by issuing a PUT request to the /connectors/" + connName + "/stop endpoint"), null);
-            return;
+            cb.onCompletion(new BadRequestException("Connectors must be in the STOPPED state before their offsets can be modified externally. " +
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java]: Nit: "modified externally" may not be very clear to users. I think "altered" captures the general concept well enough even if they're doing a reset, but if you'd like to be more precise, maybe we can say "altered or reset"?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java [hunk 3]
```diff
@@ -1587,10 +1595,11 @@ private boolean alterConnectorOffsetsChecks(String connName, Callback<Message> c
...
         // we can be sure that the tasks have at least been attempted to be stopped (or cancelled if they took too long to stop).
         // Zombie tasks are handled by a round of zombie fencing for exactly once source connectors. Zombie sink tasks are handled
-        // naturally because requests to alter consumer group offsets will fail if there are still active members in the group.
+        // naturally because requests to alter consumer group offsets / delete consumer groups will fail if there are still active members
+        // in the group.
         if (configState.targetState(connName) != TargetState.STOPPED || configState.taskCount(connName) != 0) {
-            callback.onCompletion(new BadRequestException("Connectors must be in the STOPPED state before their offsets can be altered. This " +
-                    "can be done for the specified connector by issuing a PUT request to the /connectors/" + connName + "/stop endpoint"), null);
+            callback.onCompletion(new BadRequestException("Connectors must be in the STOPPED state before their offsets can be modified externally. " +
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java]: Same thought RE "modified externally" (though seeing how "modified"/"modify" is used in some other log and exception messages in this class, I think just "modified" could work here too).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 4]
```diff
@@ -1268,39 +1270,55 @@ public void alterConnectorOffsets(String connName, Map<String, String> connector
...
+    }
+
+    /**
+     * Reset a connector's offsets.
+     *
+     * @param connName the name of the connector whose offsets are to be reset
+     * @param connectorConfig the connector's configurations
+     * @param cb callback to invoke upon completion
+     */
+    public void resetConnectorOffsets(String connName, Map<String, String> connectorConfig, Callback<Message> cb) {
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Do we need separate methods for resetting and altering offsets, or can we borrow the `modifyConnectorOffsets` strategy from the `AbstractHerder` and accept a possibly-null `offsets` parameter to distinguish between the two operations?

If we're worried about accidentally introducing a nasty bug where an empty-bodied alter request causes an unintentional reset, we can add an integration test for that case.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 5]
```diff
@@ -1320,89 +1338,188 @@ void alterSinkConnectorOffsets(String connName, Connector connector, Map<String,
...
-
-                    if (!offsetsToAlter.isEmpty()) {
-                        log.debug("Committing the following consumer group offsets using an admin client for sink connector {}: {}.",
-                                connName, offsetsToAlter);
-                        AlterConsumerGroupOffsetsOptions alterConsumerGroupOffsetsOptions = new AlterConsumerGroupOffsetsOptions().timeoutMs(
+                    Map<TopicPartition, Long> offsetsToWrite;
+                    if (isReset) {
+                        offsetsToWrite = new HashMap<>();
+                        ListConsumerGroupOffsetsOptions listConsumerGroupOffsetsOptions = new ListConsumerGroupOffsetsOptions().timeoutMs(
                                 (int) ConnectResource.DEFAULT_REST_REQUEST_TIMEOUT_MS);
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Can we use a `Timer` to limit the total runtime of this method to 90 seconds, instead of using a 90-second timeout for each admin client request?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 6]
```diff
@@ -1320,89 +1338,188 @@ void alterSinkConnectorOffsets(String connName, Connector connector, Map<String,
...
 
     /**
-     * Alter a source connector's offsets.
+     * Alter a sink connector's consumer group offsets. This is done via calls to {@link Admin#alterConsumerGroupOffsets}
+     * and / or {@link Admin#deleteConsumerGroupOffsets}.
      *
-     * @param connName the name of the source connector whose offsets are to be altered
+     * @param connName the name of the sink connector whose offsets are to be altered
+     * @param groupId the sink connector's consumer group ID
+     * @param admin the {@link Admin admin client} to be used for altering the consumer group offsets; should be closed after use
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: We close the admin client in this method, right? I think we can be clearer about that:
```suggestion
     * @param admin the {@link Admin admin client} to be used for altering the consumer group offsets; will be closed after use
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 7]
```diff
@@ -1320,89 +1338,188 @@ void alterSinkConnectorOffsets(String connName, Connector connector, Map<String,
...
+                    cb.onCompletion(new ConnectException("Failed to alter consumer group offsets for connector " + connName, error), null);
+                }
+            } else {
+                completeAlterOffsetsCallback(alterOffsetsResult, false, cb);
+            }
+        }).whenComplete((ignored, ignoredError) -> {
+            // errors originating from the original future are handled in the prior whenComplete invocation which isn't expected to throw
+            // an exception itself, and we can thus ignore the error here
+            Utils.closeQuietly(admin, "Offset alter admin for sink connector " + connName);
+        });
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: It looks like this is just pulled out from the [existing `alterSinkConnectorOffsets` method](https://github.com/apache/kafka/blob/9cfc4b9373bf887ec2fb95bc607f4ebd1c8613c0/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java#L1323-L1383) (which is fine); let me know if there are any other changes that I've missed?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 8]
```diff
@@ -1320,89 +1338,188 @@ void alterSinkConnectorOffsets(String connName, Connector connector, Map<String,
...
+    }
+
+    /**
+     * Reset a sink connector's consumer group offsets. This is done by deleting the consumer group via a call to
+     * {@link Admin#deleteConsumerGroups}
+     *
+     * @param connName the name of the sink connector whose offsets are to be reset
+     * @param groupId the sink connector's consumer group ID
+     * @param admin the {@link Admin admin client} to be used for resetting the consumer group offsets; should be closed after use
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]:
```suggestion
     * @param admin the {@link Admin admin client} to be used for resetting the consumer group offsets; will be closed after use
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 9]
```diff
@@ -1460,30 +1602,31 @@ void alterSourceConnectorOffsets(String connName, Connector connector, Map<Strin
...
+                cb.onCompletion(ConnectUtils.maybeWrap(t, "Failed to modify offsets for source connector " + connName), null);
             } finally {
-                Utils.closeQuietly(offsetStore::stop, "Offset store for offset alter request for connector " + connName);
+                Utils.closeQuietly(offsetStore::stop, "Offset store for offset modification request for connector " + connName);
             }
         }));
     }
 
-    private void completeAlterOffsetsCallback(boolean alterOffsetsResult, Callback<Message> cb) {
+    private void completeAlterOffsetsCallback(boolean alterOffsetsResult, boolean isReset, Callback<Message> cb) {
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Should we rename this to `completeModifyOffsetsCallback` to match the other naming changes we've made in this class?

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java [hunk 10]
```diff
@@ -574,6 +574,199 @@ public void alterAndVerifySourceConnectorOffsets(EmbeddedConnectCluster connect,
...
+    @Test
+    public void testResetSourceConnectorOffsetsExactlyOnceSupportEnabledAndDifferentKafkaClusterTargeted() throws Exception {
+        Properties brokerProps = new Properties();
+        brokerProps.put("transaction.state.log.replication.factor", "1");
+        brokerProps.put("transaction.state.log.min.isr", "1");
+        EmbeddedKafkaCluster connectorTargetedKafkaCluster = new EmbeddedKafkaCluster(1, new Properties());
+        workerProps.put(DistributedConfig.EXACTLY_ONCE_SOURCE_SUPPORT_CONFIG, "enabled");
+
+        // This embedded Connect cluster will internally spin up its own embedded Kafka cluster
+        EmbeddedConnectCluster exactlyOnceSupportEnabledConnectCluster = new EmbeddedConnectCluster.Builder()
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java]: Do we need to close this when we're done with it?

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java [hunk 11]
```diff
@@ -574,6 +574,199 @@ public void alterAndVerifySourceConnectorOffsets(EmbeddedConnectCluster connect,
...
+                .name("connect-cluster")
+                .brokerProps(brokerProps)
+                .numWorkers(NUM_WORKERS)
+                .workerProps(workerProps)
+                .build();
+
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java:805]: Nit: extra newline
```suggestion
```

**Review by C0urante — COMMENTED** [review]
This looks great, thanks Yash! I've made it through all of the functional changes and the integration tests; will review the unit tests in the next pass.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 1 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Good call--I agree that deleting the consumer group is better here in order to maintain consistency 👍

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java (hunk 3 above)
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java]: I think just "modified" should be clear enough to users since this is synchronously surfaced to the user via the alter / reset offsets REST API response 👍

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 6 above)
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Good catch, thanks. I think this was leftover from a previous refactor (earlier it was the caller's responsibility to close the admin client).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 7 above)
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Yeah, no changes here. I'm surprised that the GitHub diff UI isn't smarter about things like this 🤷‍♂️

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 4 above)
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: > If we're worried about accidentally introducing a nasty bug where an empty-bodied alter request causes an unintentional reset, we can add an integration test for that case.

Yeah, this was exactly my worry and the reason why I'd kept them separated. Based on your feedback, I've added a new integration test and also moved the second level check to `AbstractHerder::alterConnectorOffsets` (so that we can consolidate the two methods in the `Worker`). While we could in theory do a similar consolidation for the `Herder` methods, I think it's probably a better idea to have cleaner and more well-defined interface methods there.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 1 above)
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: I've made this change but it also brought me to the realization that we will have undesirable behavior if the connector is targeting a Kafka cluster different from the Connect cluster's backing Kafka cluster and the user has configured the consumer overrides appropriately for their connector, but not the admin overrides (something we also discussed previously [here](https://github.com/apache/kafka/pull/13434#discussion_r1144415671)). 

In the above case, if a user attempts to reset their sink connector's offsets via the `DELETE /connectors/{connector}/offsets` endpoint, the following will occur:

1. We list the consumer group offsets via `Admin::listConsumerGroupOffsets` which returns an empty partition offsets map for the sink connector's consumer group ID (it exists on a different Kafka cluster to the one that the admin client is connecting to).
2. We call `SinkConnector::alterOffsets` with an empty offsets map which could cause the sink connector to propagate the offsets reset related changes to the sink system.
3. We attempt to delete the consumer group via `Admin::deleteConsumerGroups` which returns `GroupIdNotFoundException` which we essentially swallow in order to keep offsets reset operations idempotent and return a success message to the user (even though the real consumer group for the sink connector on the other Kafka cluster hasn't been deleted).

This will occur if the connector's admin overrides are missing OR the admin overrides are deliberately configured to target a Kafka cluster different from the consumer overrides (although like you pointed out in the other linked thread, this doesn't seem like a valid use case that we'd even want to support). 

I guess we'd want to pursue the approach you suggested where we'd configure the admin client with a combination of the connector's admin overrides and consumer overrides? 

Another option could potentially be to somehow verify that the `admin.override.bootstrap.servers` in the connector's config / `admin.bootstrap.servers` in the worker config / `bootstrap.servers` in the worker config (in order of preference) correspond to the same Kafka cluster as `consumer.override.bootstrap.servers` in the connector's config / `consumer.bootstrap.servers` in the worker config / `bootstrap.servers` in the worker config (in order of preference) and fail the request if we are able to reliably determine that they aren't pointing to the same Kafka clusters? I'm not sure that this is a feasible approach however.

Yet another option could be to remove the idempotency guarantee from the `DELETE /connectors/{connector}/offsets` endpoint and if we encounter a `GroupIdNotFoundException` from `Admin::deleteConsumerGroups`, return an error message to the user indicating that either the offsets have already been reset previously or else they might need to check their connector's admin overrides (this does seem fairly ugly though).

Edit: A more elegant way might be to switch the offset reset mechanism from deleting the consumer group to deleting the offsets for all topic partitions via `Admin::deleteConsumerGroupOffsets` (similar to what we do for the `PATCH /connectors/{connector}/offsets` endpoint when the offset for a partition is specified as `null`). This way we could explicitly check for existence of the sink connector's consumer group prior to listing its offsets and fail requests if the consumer group doesn't exist (the minor down-side is that this will require an additional admin client request).

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: @yashmayya this PR is already fairly complex. Do you believe it's necessary to block on handling this kind of connector misconfiguration or is it acceptable to leave that for a follow-up item, like we did in https://github.com/apache/kafka/pull/13434?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 4 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: LGTM 👍

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java [hunk 12]
```diff
@@ -574,6 +574,274 @@ public void alterAndVerifySourceConnectorOffsets(EmbeddedConnectCluster connect,
...
+    @Test
+    public void testResetSourceConnectorOffsetsExactlyOnceSupportEnabled() throws Exception {
+        Properties brokerProps = new Properties();
+        brokerProps.put("transaction.state.log.replication.factor", "1");
+        brokerProps.put("transaction.state.log.min.isr", "1");
+        workerProps.put(DistributedConfig.EXACTLY_ONCE_SOURCE_SUPPORT_CONFIG, "enabled");
+
+        // This embedded Connect cluster will internally spin up its own embedded Kafka cluster
+        EmbeddedConnectCluster exactlyOnceSupportEnabledConnectCluster = new EmbeddedConnectCluster.Builder()
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java]: One more thought: in this case (and one other), we don't use the embedded Connect cluster created in the `@Before` method for this test suite. This is a little inefficient and can bloat test runtime.

We can follow the same approach as the [ConnectWorkerIntegrationTest](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java), where we create an `EmbeddedConnectCluster.Builder` instance in the `@Before`-annotated method for the test suite, but only build and start the actual Connect cluster in each test case. It's a little more verbose but IMO it's worth the tradeoff since we gain more in decreased test time than we lose in extra lines of code.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java [hunk 13]
```diff
@@ -1819,16 +1825,25 @@ public void testGetSinkConnectorOffsetsAdminClientAsynchronousError() {
...
+            adminFuture.completeExceptionally(e);
+        } else {
+            adminFuture.complete(consumerGroupOffsets);
+        }
+        when(result.partitionsToOffsetAndMetadata()).thenAnswer(invocation -> {
+            if (time == null) {
+                return adminFuture;
+            }
+            time.sleep(delayMs);
+            return adminFuture;
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java]: Nit: can simplify this
```suggestion
            if (time != null)
                time.sleep(delayMs);
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java [hunk 14]
```diff
@@ -2240,6 +2255,193 @@ public void testAlterOffsetsSinkConnectorSynchronousError() throws Exception {
...
+        worker.modifySinkConnectorOffsets(CONNECTOR_ID, sinkConnector, connectorProps, null,
+                Thread.currentThread().getContextClassLoader(), cb);
+        assertEquals("The offsets for this connector have been reset successfully", cb.get(1000, TimeUnit.MILLISECONDS).message());
+
+        ArgumentCaptor<DeleteConsumerGroupsOptions> deleteConsumerGroupsOptionsArgumentCaptor = ArgumentCaptor.forClass(DeleteConsumerGroupsOptions.class);
+        verify(admin).deleteConsumerGroups(anyCollection(), deleteConsumerGroupsOptionsArgumentCaptor.capture());
+        // Expect the call to Admin::deleteConsumerGroups to have a timeout value equal to the overall timeout value of DEFAULT_REST_REQUEST_TIMEOUT_MS
+        // minus the delay introduced in the call to Admin::listConsumerGroupOffsets (2000 ms) and the delay introduced in the call to
+        // SinkConnector::alterOffsets (3000 ms)
+        assertEquals((int) ConnectResource.DEFAULT_REST_REQUEST_TIMEOUT_MS - 2000L - 3000L,
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java:2340]: Very nice!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 15]
```diff
@@ -1320,89 +1317,192 @@ void alterSinkConnectorOffsets(String connName, Connector connector, Map<String,
...
+                    boolean alterOffsetsResult;
+                    try {
+                        alterOffsetsResult = ((SinkConnector) connector).alterOffsets(connectorConfig, offsetsToWrite);
+                    } catch (UnsupportedOperationException e) {
+                        log.error("Failed to modify offsets for connector {} because it doesn't support external modification of offsets",
+                                connName, e);
+                        throw new ConnectException("Failed to modify offsets for connector " + connName + " because it doesn't support external " +
+                                "modification of offsets", e);
                     }
+                    updateTimerAndCheckExpiry(timer, "Timed out while calling the 'alterOffsets' method for sink connector " + connName);
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:1353]: Should we add a note to `SourceConnector::alterOffsets` and `SinkConnector::alterOffsets` that this method may be invoked multiple times with the same arguments if an external operation (writing to the offsets topic, deleting a consumer group, etc.) fails? That way, developers can know to silently ignore requests to reset offsets that don't appear to be present based on the connector's own external tracking logic.

**Review by C0urante — COMMENTED** [review]
Looking great! Some minor comments and then this should be ready to merge (pending CI build). Thanks Yash!

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java (hunk 12 above)
Comment (yashmayya) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java]: Hah, I was considering doing the same thing earlier but discarded it thinking that it'd add more noise for review (since we'd touch existing test cases too). Thanks for the push in the right direction (I agree that adding a couple of lines of code to tests here and there is worth it for reducing test runtime bloat); I've made this change 👍

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 1 above)
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Yeah, that makes sense. I've filed this follow-up Jira ticket so that we can re-visit if necessary - https://issues.apache.org/jira/browse/KAFKA-15113

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 15 above)
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:1353]: Hm good point, we really don't want to get into a situation where let's say a call to `SinkConnector::alterOffsets` succeeds but the call to delete the consumer group fails and then any subsequent external requests to delete the connector's offsets fails because the `SinkConnector::alterOffsets` method bails out..

On the other hand, I'm also concerned about making the Javadoc for those methods overly complicated. Furthermore, since we'll be calling `alterOffsets` on separate `Connector` instances, it's not like they can track state between multiple calls. Any suggestions on the wording so that we can maintain the right balance? Maybe we could add a hint that offsets that appear non-existent should be handled gracefully if possible (this also covers the case where an alter offsets request is used to skip over some data for sink connectors)?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 1 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Awesome, thanks!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 15 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:1353]: > Maybe we could add a hint that offsets that appear non-existent should be handled gracefully if possible (this also covers the case where an alter offsets request is used to skip over some data for sink connectors)?

I like this, but I wish we could be a little more explicit. I think ideally we want connectors to respond with a no-op for resets of partitions that don't appear to exist, and for non-resets (assuming they do external offset tracking at all) either store that partition/offset pair externally, or throw an exception to indicate that the operation cannot be performed.

Obviously that's a bit wordy on its own and I agree that we shouldn't get too wordy in the Javadoc... feel free to push the best you can come up with and if necessary we can either workshop it further in this PR or a follow-up.

**Review by C0urante — APPROVED** [review]
💯 LGTM, thanks Yash!

**Comment by C0urante** [comment]
Test failures appear unrelated; merging...
